network/simulation/simulation.go

```diff
@@ -71,13 +80,49 @@ type ServiceFunc func(ctx *adapters.ServiceContext, bucket *sync.Map) (s node.Se
 // every ServiceFunc must return a node.Service of the unique type.
 // This restriction is required by node.Node.Start() function
 // which is used to start node.Service returned by ServiceFunc.
-func New(services map[string]ServiceFunc) (s *Simulation) {
+func NewInProc(services map[string]ServiceFunc) (s *Simulation) {
+	s = &Simulation{
+		buckets:           make(map[enode.ID]*sync.Map),
+		done:              make(chan struct{}),
+		neighbourhoodSize: network.NewKadParams().NeighbourhoodSize,
+		typ:               SimulationTypeInproc,
+	}
+
+	adapterServices := s.addServices(services)
+
+	s.Net = simulations.NewNetwork(
+		adapters.NewTCPAdapter(adapterServices),
+		&simulations.NetworkConfig{ID: "0"},
+	)
+
+	return s
+}
+
+// NewExec does the same as New but lets the caller specify the adapter to use
+func NewExec(services map[string]ServiceFunc) (s *Simulation, err error) {
 	s = &Simulation{
 		buckets:           make(map[enode.ID]*sync.Map),
 		done:              make(chan struct{}),
 		neighbourhoodSize: network.NewKadParams().NeighbourhoodSize,
+		typ:               SimulationTypeExec,
 	}
 
+	adapterServices := s.addServices(services)
```

Comment:
This is still confusing (but better than before), because `s.addServices` is not only returning `adapterServices`, but also modifying `s`.

Maybe it'd be better is line 111 was part of `s.addServices` so that you don't have to return the `adapterServices` and rely on `s.addServices` to add them to the `simulation` and register them.

Reply:
Thing is we don't need to `RegisterServices` there for the inproc node.